source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1289 @@
+
+Add the `map` attribute set with the `zoom` option into the top-level `options` declaration, like so:
+
+```diff
+# default.nix
+     requestParams = lib.mkOption {
+       type = lib.types.listOf lib.types.str;
+     };
++
++    map = {
++      zoom = lib.mkOption {
++        type = lib.types.nullOr lib.types.int;
++      };
++    };
+   };
+```
+
+To make use of this, use the `mkIf <condition> <definition>` function, which only adds the definition if the condition evaluates to `true`.
+Make the following additions to the `requestParams` list in the `config` block:
+
+```diff
+# default.nix
+     requestParams = [
+       "size=640x640"
+       "scale=2"
++      (lib.mkIf (config.map.zoom != null)
++        "zoom=${toString config.map.zoom}")
+     ];
+   };
+```
+
+This will will only add a `zoom` parameter to the script call if the value is non-null.
+
+## Default values
+
+Let's say that in our application we want to have a different default behavior that sets the the zoom level to `2`, such that automatic zoom has to be enabled explicitly.
+
+This can be done with the `default` argument to [`mkOption`](https://github.com/NixOS/nixpkgs/blob/master/lib/options.nix).
+Its value will be used if the value of the option declaring it is not specified otherwise.
+
+Add the corresponding line:
+
+```diff
+# default.nix
+     map = {
+       zoom = lib.mkOption {
+         type = lib.types.nullOr lib.types.int;
++        default = 2;
+       };
+     };
+   };
+```
+
+## Centering the Map
+
+You have now declared options controlling the map dimensions and zoom level, but have not provided a way to specify where the map should be centered.
+
+Add the `center` option now, possibly with your own location as default value:
+
+```diff
+# default.nix
+         type = lib.types.nullOr lib.types.int;
+         default = 2;
+       };
++
++      center = lib.mkOption {
++        type = lib.types.nullOr lib.types.str;
++        default = "switzerland";
++      };
+     };
+   };
+```
+
+To implement this behavior, you will use the {download}`geocode <files/geocode>` utility, which turns location names into coordinates.
+There are multiple ways of making a new package accessible, but as an exercise, you will add it as an option in the module system.
+
+First, add a new option to accommodate the package:
+
+
+```diff
+# default.nix
+   options = {
+     scripts.output = lib.mkOption {
+       type = lib.types.package;
+     };
++
++    scripts.geocode = lib.mkOption {
++      type = lib.types.package;
++    };
+```
+
+Then define the value for that option where you make the raw script reproducible by wrapping a call to it in `writeShellApplication`:
+
+```diff
+# default.nix
+   config = {
++    scripts.geocode = pkgs.writeShellApplication {
++      name = "geocode";
++      runtimeInputs = with pkgs; [ curl jq ];
++      text = "exec ${./geocode}";
```

Comment:
Might as well fetch this (and `map`) using Nix, so there's no need for a manual download :)